### ISSUE TITLE
Provide a way to not open a link in the PWA, in the moment

### ISSUE THREAD

**Issue by PieterjanDeClippel** [w3c/manifest#989]
Many web apps have a login. This login page is often opened in a new window, in order not to interrupt the already running web application. However when a web app is installed as PWA, `window.open()` will always open the link in the PWA.

The usual flow in single-page applications is:
- Subscribe to the `window.onmessage` event
- Use `window.open()` to navigate to the URL hosting the login page (for eg. Microsoft OAuth)
- Host an empty html page with only javascript, using the messaging api to notify the `opener`.
- Let the OAuth flow use this page as redirect_url.

But since `window.open()` always opens the PWA, the external-login flow is interrupted.
The bottom line is: a developer MUST BE allowed to have window.open open the url in a new window and prevent it from opening the PWA.

**Comment by alancutter** [comment]
What is the interruption in "external-login flow is interrupted"?

**Comment by PieterjanDeClippel** [comment]
Thanks for the reply.

![ezgif-2-951fbd944edc](https://user-images.githubusercontent.com/9629574/126951208-bc46bf7b-e29f-4c83-9747-ca503da66015.gif)

After clicking the "Login using xxx" button, I'm executing the following:

    this.authWindow = window.open(`${this.baseUrl}/web/${this.apiVersion}/Account/${this.action}/${medium}/${this.platform}`, '_blank', 'width=600,height=400');
    // For example
    // this.authWindow = window.open(`https://example.com/web/v3/Account/connect/web/Microsoft`, '_blank', 'width=600,height=400');

Since **https://example.com** is the root of my PWA, you can see that `window.open` is opening my PWA instead of just a new tab. After signing in into the Microsoft/Facebook/... login, a script is executed:

```
<!DOCTYPE html>
<html lang="en">
<head>
	<title>Bezig met verwerken...</title>
	<script type="text/javascript">
        var message = {
            status: @( (int)Model.Status ),
            medium: "@Model.Medium",
            platform: "@Model.Platform",
            error: "@Model.Error",
            errorDescription: "@Model.ErrorDescription"
        };
        window.opener.postMessage(JSON.stringify(message), "https://example.com");
	</script>
</head>
<body>
</body>
</html>
```

But since this url has been opened in the PWA instead of a browser window, the message doesn't arrive in the opener's window.

> opener -> opens `https://example.com/signin-microsoft` -> PWA is being launched -> Link with opener is broken.

**Comment by PieterjanDeClippel** [comment]
I should be able to tell `window.open` to open the url in the browser, and not in the PWA. As stated in [this SO thread](https://stackoverflow.com/questions/68518730/pwa-prevent-urls-from-opening-in-the-pwa?noredirect=1#comment121092911_68518730), I already tried everything.

- Exclude from navigationUrls: `"navigationUrls": [ "!/web/v3/Account/connect/" ]`
- Use target: `window.open(url, '_system')`
- Use target: `window.open(url, '_blank')`
- Use ngsw-bypass: `window.open(url + '?ngsw-bypass=true')`

But still the links open in the PWA. A year ago I worked around this by creating a subdomain hosting the MVC endpoints for the external logins, and that works pretty well.
But now I ran into trouble while implementing 2-factor authentication, because the calls to `ExternalLoginSigninAsync` and `TwoFactorAuthenticatorSignInAsync` have to be executed from the same domain.

**Comment by PieterjanDeClippel** [comment]
And I've also already read about [Declarative Link Capturing](https://web.dev/declarative-link-capturing/), but it seems that this does not cover the case of forcing `window.open` to open in the browser and not the PWA.

**Comment by tomayac** [comment]
> And I've also already read about [Declarative Link Capturing](https://web.dev/declarative-link-capturing/), but it seems that this does not cover the case of forcing `window.open` to open in the browser and not the PWA.

Yes, that's correct: _"This proposal does not apply if the user is already within the navigation scope (for instance, if the user has a browser tab open that is within scope, and clicks an internal link)."_—[article](https://web.dev/declarative-link-capturing/#:~:text=This%20proposal%20does%20not%20apply%20if%20the%20user%20is%20already%20within%20the%20navigation%20scope%20(for%20instance%2C%20if%20the%20user%20has%20a%20browser%20tab%20open%20that%20is%20within%20scope%2C%20and%20clicks%20an%20internal%20link).)

**Comment by tomayac** [comment]
Adding the link to https://github.com/WICG/pwa-url-handler/issues/28.

**Comment by LuHuangMSFT** [comment]
I suspect this is an Android WebAPK behavior.

> When a Progressive Web App is installed on Android, it will register a set of intent filters for all URLs within the scope of the app. When a user clicks on a link that is within the scope of the app, the app will be opened, rather than opening within a browser tab.

https://developers.google.com/web/fundamentals/integration/webapks#android_intent_filters

**Comment by PieterjanDeClippel** [comment]
I did a little test, and installed the PWA for all of Chrome, Edge and Firefox (only on android, since the problem is only on android).
The window.open behavior is as follows:

Chrome: Opens in the chrome-installed-PWA
Edge: Opens in the edge-browser
Firefox: Opens in the firefox-browser

So it seems like, on android, the different browsers have a different implementation of the W3-drafts... Should this issue be transferred to the Chromium team then?

**Comment by alancutter** [comment]
I think filing a crbug.com is the way forward here. This sounds like an incompatibility with how Android Chrome handles web app link capturing especially given Firefox appears to handle it fine.

**Comment by PieterjanDeClippel** [comment]
Adding backlink to [Chromium bugtracker](https://bugs.chromium.org/p/chromium/issues/detail?id=1236536)

**Comment by PieterjanDeClippel** [comment]
[This block of code][2] appears to determine whether the PWA should be launched or a new tab should be opened.

Call stack:

- [WebappLauncherActivity.shouldLaunchWebapp](https://github.com/chromium/chromium/blob/master/chrome/android/java/src/org/chromium/chrome/browser/webapps/WebappLauncherActivity.java#L214)
- [WebApkValidator.canWebApkHandleUrl](https://github.com/chromium/chromium/blob/master/components/webapk/android/libs/client/src/org/chromium/components/webapk/lib/client/WebApkValidator.java#L120)
- [WebApkValidator.resolveInfosForUrlAndOptionalPackage](https://github.com/chromium/chromium/blob/master/components/webapk/android/libs/client/src/org/chromium/components/webapk/lib/client/WebApkValidator.java#L149) should return `false` for the urls you don't want opened in the PWA

**Comment by PieterjanDeClippel** [comment]
I hope that perhaps the draft can be updated, maybe stating something like this:

> When a url-open-request is to be processed, the opener must verify if the origin's hoststring matches the hoststring of the requested url. If this is the case, the url must be opened as a new tab within the same browser.
> If the hoststrings do not match, the opener can call the operating system's handler for urls, which in its turn verifies if any native applications can handle the url.

**Comment by marcoscaceres** [comment]
Shouldn't the user agent just respect `window.open()` tho? That's literally saying open a popup, so it should do that as normal.

**Comment by PieterjanDeClippel** [comment]
It seems that Firefox For Android doesn't even attempt to open the urls natively. Although I can understand that not all web browsers would opt-in into native behavior, there appears to be a huge inconsistency in how url-open requests are handled by browsers.

**Comment by marcoscaceres** [comment]
@PieterjanDeClippel understood. 

I think we to distill this down to a simple test case that we can then send around to browser vendors. 

I made a super simple PWA test case:
https://github.com/marcoscaceres/playground/tree/gh-pages/popup

You can try it here:
https://marcoscaceres.github.io/playground/popup/

We can use that if you'd like to file bugs on each individual browser vendor. 

(FWIW, Safari seems to do the right thing cross-origin popups at least - it opens an overlay window/browser)  

Note that this is still somewhat outside the scope of this specification... although we could provide some guidance around opening auxiliary browsing contexts.

**Comment by barhemo** [comment]
+1

**Comment by verekia** [comment]
Maybe another use case could revive this conversation. I am making a [web game portal](https://webgamer.io/), which is basically simply a list of games that open in a new tab when clicked. The portal itself can be installed as a PWA. I want games to open in the users' default browser, not inside the PWA, because it is a more familiar and full-featured experience for the user.

On desktop (Chrome, Mac), clicking on a game in the PWA opens it in the default browser, which is exactly what I want!

On Android, when opening a game from the PWA portal, it opens it in the in-PWA web view, which is a pretty bad user experience because the user cannot rotate to landscape mode in that view, and cannot use regular browser features in general. This makes the experience of using the PWA portal worse than using the portal in the browser. That's unfortunate.

On iOS, launching the PWA (after adding it to the home screen manually) is a bit better, because the web view allows rotating to landscape mode, but still, I would prefer launching the game in the user's default browser.

My `scope` is left blank. I tried different `scope` variations and it didn't change this behavior.

**Comment by verekia** [comment]
@tomayac Thank you for sharing this tester app and for your insights!

For what it's worth, on my Pixel 5, Android 13, Chrome 109, none of the options of the tester app open in the main browser. The 4 options you listed above open in the in-PWA browser.

**Comment by tomayac** [comment]
> @tomayac Thank you for sharing this tester app and for your insights!
> 
> For what it's worth, on my Pixel 5, Android 13, Chrome 109, none of the options of the tester app open in the main browser. The 4 options you listed above open in the in-PWA browser.

Yes, that’s the current behavior on Android. We agree it’s not always the ideal behavior.

**Comment by kurwaBober** [comment]
maybe there are any updates on this problem?

**Comment by piotr-cz** [comment]
We use PWA with specific route for an OAuth login and we are unable to `postMessage` back as the `window.opener` is null.

**Comment by monstermac77** [comment]
We're hitting this issue as well. We have to declare the scope of the PWA to be our entire domain, but we have certain links that should not be opened in the PWA (in fact many of users seem to be getting into a circular state based on some support and Google Play reviews, where they tap links inside the PWA that open in the PWA even though they should open in the browser, and then on those pages they try to get back to the app which opens the Play Store, and then they find the loop where opening the app from the Play Store just takes them back to the page they were on just before the Play Store). Confusing, I know, but this would be entirely avoided if there was a directive that could be applied, or "target=_blank" was respected, so that PWAs can better replicate native apps (e.g. we haven't heard from any of our native iOS users about this, because the expected behavior occurs where the links open in Safari).

**Comment by a-type** [comment]
I was hoping to make a suite of separate PWAs which are all tied to the same user account system. The idea was that I could include an affordance for switching between apps, similar to Google's app selector, and those links to other apps would open their respective PWA. Likewise you could open a link to manage your account which would take you back to the browser / "hub" site.

I'm gathering that this isn't feasible on mobile today. If I include a link to another one of my apps within an app, the second app will only open in a webview. That's going to be prohibitively confusing.

I'm not seeing a way to execute on this concept given the functionality available today, which is disappointing. By and large the limitations of PWAs have been avoidable, but once I hit one there's basically no getting by it.

**Comment by Knud13** [comment]
I'm also having this issue.
We use the PayPal Buttons SDK on our site which works fine... until run on Android in the PWA.
By default 3rd party cookies are blocked (which I'm fine with) but when our customers try and make a payment the PayPal sign in goes into a loop trying to verify identity.
Easy I'll just make the payment page open in the browser with target="_blank" but this doesn't work. ahhh.
Is there a way to make one page open in the browser and not in the PWA frame?

**Comment by piotr-cz** [comment]
I wonder if this issue might be mitigated with [Tabbed web apps](https://chromestatus.com/feature/5128143454076928), where `new_tab_button` would point to OAuth login page

**Comment by monstermac77** [comment]
> I wonder if this issue might be mitigated with [Tabbed web apps](https://chromestatus.com/feature/5128143454076928), where `new_tab_button` would point to OAuth login page

Definitely not preferred but you're right this would be a really good stopgap until they actually provide us a way to handle this!

**Comment by piotr-cz** [comment]
Unfortunately my workaround above won't work for OAuth, as
> The url member is a string that represents a URL relative to the [manifest URL](https://www.w3.org/TR/appmanifest/#dfn-manifest-url) that is [within scope](https://www.w3.org/TR/appmanifest/#dfn-within-scope-0) of a [processed manifest](https://www.w3.org/TR/appmanifest/#dfn-processed-manifest).

Reference: https://wicg.github.io/manifest-incubations/#ref-for-dfn-new_tab_button-2

**Comment by a-type** [comment]
I'm realizing that the use case I'm looking for (opening the user's default browser / pwa host browser when clicking a link inside the PWA) is different from the original problem cited in this issue (opening a link in the user's browser opens the PWA because it matches the scope).

Given this issue isn't getting any attention, I wonder if it would be worth diverging these two use cases. I think some other commenters are also in my situation. Just want to keep things disambiguated.

For my problem, a few solutions come to mind:
- Just have PWAs interpret `_blank` as a link for the user's browser (probably not viable as it breaks with existing behavior)
- Add a manifest option to 'allow list' particular origins for external opening instead of in-pwa frame view
- Add a new attribute or new attribute value for anchor tags which hints to PWAs to open in the user's browser instead of in-pwa.

**Comment by PieterjanDeClippel** [comment]
https://stackoverflow.com/a/59319252/8941307

**Comment by PieterjanDeClippel** [comment]
3 years on.

Done waiting for a draft.

Did it [like this](https://github.com/PieterjanDeClippel/pwatest/blob/master/src/app/app.component.ts#L17)

**Comment by a-type** [comment]
@PieterjanDeClippel Not working for me on Android 14, Chrome 124.0.6367.179. Still opens in a frame.

**Comment by dmurph** [comment]
I don't have any info about changing the behavior on Android here, but I have been working on getting (user controllable) link capturing to work on desktop platforms in Chromium, and our current plan is to only allow links to be captured into app launches if the link is creating a new 'top level transversable' [spec](https://html.spec.whatwg.org/multipage/document-sequences.html#creating-a-new-top-level-traversable), which basically means when there is supposed to be a new frame AND there isn't a shared browser context group with the opener (AKA no `window.opener`, no `window.frames`, etc). This makes sure the web platform stuff doesn't break, and from all of the examples we've looked at should match user expectations.

Will this possibly affect Android? I don't know. But I'll poke around to see how engrained that behavior is on the Android platform, and if it's feasible to change in the future. No promises though, this seems like a core thing on the Android platform (but I really have no idea).

**Comment by monstermac77** [comment]
@dmurph Thanks so much for your attention on this, Daniel!

**Comment by a-type** [comment]
@dmurph Speaking as someone lacking technical knowledge of spec and implementations here, is this interpretation of practical outcomes correct?

- A link in a PWA with `rel="noopener"` and/or `target="_blank"` will open in another app - be it browser or other app registered to handle the link, like another PWA
- A link in a PWA with `rel="opener"` will (may?) open inside the PWA as a frame overlay

That would be great, if true. Understood that your scope of work is not necessarily encompassing Android or other mobile OS, but setting some kind of standard for this functionality is a good start. It's a severe limitation today.

**Comment by dmurph** [comment]
Question about the paypal use-case - is this an issue with the paypal native app being installed? Or is this a paypal web app installed? Can you give me more context about that repro case?

**Comment by Knud13** [comment]
> Question about the paypal use-case -

Our paypal use-case...
• chrome browser 3rd party cookies Allowed:
**works as excpected** in both the browser or running out site as a PWA
• chrome browser 3rd party cookies disabled:
**works as excpected** If you have 3rd party cookies disabled in your chrome browser and use the paypal buttons the PayPal popup opens itself in a new tab and the payment process works ( this works exactly the same with PayPal app installed or not installed )
• chrome browser 3rd party cookies disabled and our site installed as a PWA
**BROKEN** Here is where the problem occurs. When using the PWA version of our site the PayPal buttons open an in page popup and the PayPal sign in process fails (without an error) and cookies messages keep showing from paypal. This fails both with PayPal app installed or not installed.
I tried making a link to our payment page with target="_blank" but this opens in a in-PWA web view and the PayPal process still fails.

I need a way to get the PayPal sign in to work within the PWA (e.g. allow the required PayPal cookies) or a way to open our payment page in the browser.

Cheers

**Comment by Moongazer** [comment]
Found this issue because we have many payment-issues using PayPal and also other payment providers, where the checkout starts inside the PWA but certain login or redirect-urls leaving the scope and opening the browser, where the session can not be continued (probably missing cookies or so)

The `scope` should be everything what we need here - if it would work like [described in docs](https://developer.mozilla.org/en-US/docs/Web/Manifest/scope#examples):
```
Finally, the following example limits navigation to a subdirectory of the current site:
"scope": "https://example.com/subdirectory/"
```

If that would work, it would be an easy solution, because links to shop or product-pages can be just out of PWA scope, to be opened automatically in the default browser, but that seems not possible yet:
- `https://example.com/subdirectory/#/login` opens in PWA because it's in the scope
- `https://example.com/product-list` clicked from inside PWA should open default browser because out of scope

After weeks of digging around, this info came up: "_Theoretically the scope should limit the PWA to only control pages within https://example.com/subdirectory/ and its subdirectories. In reality, the scope doesn't prevent navigation to out-of-scope URLs within the same origin (in this case https://example.com). The browser allows the PWA to navigate to any URL within the same origin, even if it's outside the defined scope_".

Is that true, and if so: why? That totally messes up the concept of having a scope.

**Comment by dmurph** [comment]
I created https://crbug.com/364623620 as well to track

**Comment by dmurph** [comment]
Request from that bug - does anyone have a minimal reproduction test case we can run on our end?

**Comment by glennhartmann** [comment]
@Moongazer wrote:

> After weeks of digging around, this info came up: "_Theoretically the scope should limit the PWA to only control pages within https://example.com/subdirectory/ and its subdirectories. In reality, the scope doesn't prevent navigation to out-of-scope URLs within the same origin (in this case https://example.com). The browser allows the PWA to navigate to any URL within the same origin, even if it's outside the defined scope_".
> 
> Is that true, and if so: why? That totally messes up the concept of having a scope.

That should not be the case for Android PWAs. You can try yourself with https://special-victorious-polonium.glitch.me/app - the scope is set to `/app/`, and if you click the small link labeled `nonapp`, it'll navigate to a URL outside of that scope (but on the same origin). When using the installed PWA (called a "WebAPK" on Android), you should see a URL bar at the top on out-of-scope pages, which indicates that the page is being loaded in a [Custom Tab](https://developer.chrome.com/docs/android/custom-tabs) rather that in the PWA's main context.

@Moongazer wrote:

> If that would work, it would be an easy solution, because links to shop or product-pages can be just out of PWA scope, to be opened automatically in the default browser, but that seems not possible yet:
> 
> * `https://example.com/subdirectory/#/login` opens in PWA because it's in the scope
> * `https://example.com/product-list` clicked from inside PWA should open default browser because out of scope
> 

Note that I'm not aware of any requirement that an out-of-scope URL must be loaded in the system's default browser. Quoting [the spec](https://www.w3.org/TR/appmanifest/#nav-scope):

> If the [application context](https://www.w3.org/TR/appmanifest/#dfn-application-context)'s [active document](https://html.spec.whatwg.org/multipage/document-sequences.html#nav-document)'s [URL](https://dom.spec.whatwg.org/#concept-document-url) is not [within scope](https://www.w3.org/TR/appmanifest/#dfn-within-scope-0) of the [application context](https://www.w3.org/TR/appmanifest/#dfn-application-context)'s [processed manifest](https://www.w3.org/TR/appmanifest/#dfn-processed-manifest), the user agent SHOULD show a prominent UI element indicating the [URL](https://dom.spec.whatwg.org/#concept-document-url) or at least its [origin](https://html.spec.whatwg.org/multipage/browsers.html#concept-origin), including whether it is served over a secure connection. This UI SHOULD differ from any UI used when the [URL](https://dom.spec.whatwg.org/#concept-document-url) is [within scope](https://www.w3.org/TR/appmanifest/#dfn-within-scope-0) of the [application context](https://www.w3.org/TR/appmanifest/#dfn-application-context)'s [processed manifest](https://www.w3.org/TR/appmanifest/#dfn-processed-manifest), in order to make it obvious that the user is navigating off scope.

**Comment by PieterjanDeClippel** [comment]
> Request from that bug - does anyone have a minimal reproduction test case we can run on our end?

You can try [this one](https://github.com/PieterjanDeClippel/pwatest) probably

**Comment by a-type** [comment]
It seems clear that this issue is discussing a different problem than the one I have at this point (I want to be able to open out-of-scope pages in the user's browser, not a Custom Tab / in-app experience, even if that's the affordance provided in the spec) so perhaps I'll create a new issue.

**Comment by dmurph** [comment]
So there are two issues here:

1. The original issue is that, if you installed the 'paypal' websites as a PWA, on Android we capture ALL url clicks or navigations (no matter what the target or relationship is - e.g. auxiliary frames get intercepted, same-site navigations intercepted, etc) into the installed pwa if the url is in-scope. This breaks a bunch of stuff, especially if it's something like paypal which relies on opening child frames to, say, paypal logins.
2. There is another issue of navigations IN a pwa opening in a custom tab, even if it's supposed to be target=_blank / new browser context.

I have renamed this issue to `Standardize when link clicks or navigations enter and/or leave installed PWAs` to hopefully make this more clear.

On Chrome, we have a bug for the CCT issue here: https://crbug.com/40745354 - where out of scope links that target _blank should open in browser window.

I think a big thing I'm not up to date on is if Android installed PWAs have isolated storage - I believe we wanted cookies / data to be shared with the browser, so if you log in on the browser you'll be logged in in the app, etc. If that's not the case, then the policy might have to be different here, and we might need CCTs? At least for auxiliary frames.

**Comment by rejhgadellaa** [comment]
On Android (and every other OS except iOS where Chrome can't run PWAs at all), the storage is shared between the browser and the PWA. If you log in to a site before installing the PWA, the state is shared. If you log in to the PWA and then visit the site, the state is shared.

**Comment by monstermac77** [comment]
> On Android (and every other OS except iOS where Chrome can't run PWAs at all), the storage is shared between the browser and the PWA. If you log in to a site before installing the PWA, the state is shared. If you log in to the PWA and then visit the site, the state is shared.

Just corroborating this. Also, when you install a PWA from the Google Play Store, on first launch it shows a little banner that says "This app is running in Chrome" to indicate to the user that cookies, local storage, etc. are all shared.
